Add background fetch WebIDL

youennf opened the pull request:
#### 663c322dad1acda181b1c75a52f24e61da41359a
<pre>
Add background fetch WebIDL
<a href="https://bugs.webkit.org/show_bug.cgi?id=251980">https://bugs.webkit.org/show_bug.cgi?id=251980</a>
rdar://problem/105213321

Reviewed by Chris Dumez.

Introduce WebIDL, experimental flag and minimal implementation.

Covered by rebased tests.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/background-fetch/abort.https.window-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/background-fetch/content-security-policy.https.window-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/background-fetch/fetch-uploads.https.window-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/background-fetch/fetch.https.window-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/background-fetch/get-ids.https.window-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/background-fetch/get.https.window-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/background-fetch/idlharness.https.any-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/background-fetch/idlharness.https.any.serviceworker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/background-fetch/idlharness.https.any.sharedworker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/background-fetch/idlharness.https.any.worker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/background-fetch/interfaces.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/background-fetch/match.https.window-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/background-fetch/mixed-content-and-allowed-schemes.https.window-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/background-fetch/port-blocking.https.window-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/background-fetch/update-ui.https.window-expected.txt: Added.
* Source/JavaScriptCore/runtime/IntlListFormat.h:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/fetch/FetchHeaders.h:
* Source/WebCore/Modules/fetch/FetchHeadersGuard.h: Added.
* Source/WebCore/Modules/fetch/FetchResponseBodyLoader.h: Added.
(WebCore::FetchResponseBodyLoader::takeConsumeDataCallback):
(WebCore::FetchResponseBodyLoader::hasConsumeDataCallback const):
(WebCore::FetchResponseBodyLoader::consumeDataCallback):
(WebCore::FetchResponseBodyLoader::consumeDataByChunk):
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/dom/EventNames.h:
* Source/WebCore/dom/EventNames.in:
* Source/WebCore/dom/EventTargetFactory.in:
* Source/WebCore/workers/service/background-fetch/BackgroundFetchEvent.cpp: Added.
(WebCore::BackgroundFetchEvent::create):
(WebCore::BackgroundFetchEvent::BackgroundFetchEvent):
(WebCore::BackgroundFetchEvent::~BackgroundFetchEvent):
(WebCore::BackgroundFetchEvent::registration const):
* Source/WebCore/workers/service/background-fetch/BackgroundFetchEvent.h: Added.
* Source/WebCore/workers/service/background-fetch/BackgroundFetchEvent.idl: Added.
* Source/WebCore/workers/service/background-fetch/BackgroundFetchEventInit.h: Added.
* Source/WebCore/workers/service/background-fetch/BackgroundFetchEventInit.idl: Added.
* Source/WebCore/workers/service/background-fetch/BackgroundFetchFailureReason.h: Added.
* Source/WebCore/workers/service/background-fetch/BackgroundFetchFailureReason.idl: Added.
* Source/WebCore/workers/service/background-fetch/BackgroundFetchInformation.h: Added.
(WebCore::BackgroundFetchInformation::isolatedCopy const):
(WebCore::BackgroundFetchInformation::isolatedCopy):
* Source/WebCore/workers/service/background-fetch/BackgroundFetchManager.cpp: Added.
(WebCore::BackgroundFetchManager::BackgroundFetchManager):
(WebCore::BackgroundFetchManager::~BackgroundFetchManager):
(WebCore::BackgroundFetchManager::fetch):
(WebCore::BackgroundFetchManager::get):
(WebCore::BackgroundFetchManager::getIds):
* Source/WebCore/workers/service/background-fetch/BackgroundFetchManager.h: Added.
(WebCore::BackgroundFetchManager::create):
* Source/WebCore/workers/service/background-fetch/BackgroundFetchManager.idl: Added.
* Source/WebCore/workers/service/background-fetch/BackgroundFetchOptions.h: Added.
(WebCore::BackgroundFetchOptions::BackgroundFetchOptions):
(WebCore::BackgroundFetchOptions::downloadTotal):
(WebCore::BackgroundFetchOptions::isolatedCopy const):
(WebCore::BackgroundFetchOptions::isolatedCopy):
* Source/WebCore/workers/service/background-fetch/BackgroundFetchOptions.idl: Added.
* Source/WebCore/workers/service/background-fetch/BackgroundFetchRecord.cpp: Added.
(WebCore::BackgroundFetchRecord::BackgroundFetchRecord):
(WebCore::BackgroundFetchRecord::~BackgroundFetchRecord):
(WebCore::BackgroundFetchRecord::request):
(WebCore::BackgroundFetchRecord::settleResponseReadyPromise):
* Source/WebCore/workers/service/background-fetch/BackgroundFetchRecord.h: Added.
(WebCore::BackgroundFetchRecord::create):
(WebCore::BackgroundFetchRecord::responseReady):
* Source/WebCore/workers/service/background-fetch/BackgroundFetchRecord.idl: Added.
* Source/WebCore/workers/service/background-fetch/BackgroundFetchRecordIdentifier.h: Added.
* Source/WebCore/workers/service/background-fetch/BackgroundFetchRecordInformation.h: Added.
(WebCore::BackgroundFetchRecordInformation::isolatedCopy const):
(WebCore::BackgroundFetchRecordInformation::isolatedCopy):
* Source/WebCore/workers/service/background-fetch/BackgroundFetchRegistration.cpp: Added.
(WebCore::BackgroundFetchRegistration::create):
(WebCore::BackgroundFetchRegistration::BackgroundFetchRegistration):
(WebCore::BackgroundFetchRegistration::~BackgroundFetchRegistration):
(WebCore::BackgroundFetchRegistration::abort):
(WebCore::BackgroundFetchRegistration::match):
(WebCore::BackgroundFetchRegistration::matchAll):
(WebCore::BackgroundFetchRegistration::updateInformation):
(WebCore::BackgroundFetchRegistration::activeDOMObjectName const):
(WebCore::BackgroundFetchRegistration::stop):
(WebCore::BackgroundFetchRegistration::virtualHasPendingActivity const):
* Source/WebCore/workers/service/background-fetch/BackgroundFetchRegistration.h: Added.
(WebCore::BackgroundFetchRegistration::id const):
(WebCore::BackgroundFetchRegistration::uploadTotal const):
(WebCore::BackgroundFetchRegistration::uploaded const):
(WebCore::BackgroundFetchRegistration::downloadTotal const):
(WebCore::BackgroundFetchRegistration::downloaded const):
(WebCore::BackgroundFetchRegistration::result):
(WebCore::BackgroundFetchRegistration::failureReason):
(WebCore::BackgroundFetchRegistration::recordsAvailable):
(WebCore::BackgroundFetchRegistration::registrationIdentifier const):
* Source/WebCore/workers/service/background-fetch/BackgroundFetchRegistration.idl: Added.
* Source/WebCore/workers/service/background-fetch/BackgroundFetchRequest.h: Added.
(WebCore::BackgroundFetchRequest::isolatedCopy const):
(WebCore::BackgroundFetchRequest::isolatedCopy):
* Source/WebCore/workers/service/background-fetch/BackgroundFetchResult.h: Added.
* Source/WebCore/workers/service/background-fetch/BackgroundFetchResult.idl: Added.
* Source/WebCore/workers/service/background-fetch/BackgroundFetchUIOptions.h: Added.
* Source/WebCore/workers/service/background-fetch/BackgroundFetchUIOptions.idl: Added.
* Source/WebCore/workers/service/background-fetch/BackgroundFetchUpdateUIEvent.cpp: Added.
(WebCore::BackgroundFetchUpdateUIEvent::create):
(WebCore::BackgroundFetchUpdateUIEvent::BackgroundFetchUpdateUIEvent):
(WebCore::BackgroundFetchUpdateUIEvent::~BackgroundFetchUpdateUIEvent):
(WebCore::BackgroundFetchUpdateUIEvent::updateUI):
* Source/WebCore/workers/service/background-fetch/BackgroundFetchUpdateUIEvent.h: Added.
* Source/WebCore/workers/service/background-fetch/BackgroundFetchUpdateUIEvent.idl: Added.
* Source/WebCore/workers/service/background-fetch/ImageResource.h: Added.
(WebCore::ImageResource::isolatedCopy const):
(WebCore::ImageResource::isolatedCopy):
* Source/WebCore/workers/service/background-fetch/ImageResource.idl: Added.
* Source/WebCore/workers/service/background-fetch/ServiceWorkerGlobalScope+BackgroundFetchAPI.idl: Added.
* Source/WebCore/workers/service/background-fetch/ServiceWorkerRegistration+BackgroundFetchAPI.idl: Added.
* Source/WebCore/workers/service/background-fetch/ServiceWorkerRegistrationBackgroundFetchAPI.cpp: Added.
(WebCore::ServiceWorkerRegistrationBackgroundFetchAPI::ServiceWorkerRegistrationBackgroundFetchAPI):
(WebCore::ServiceWorkerRegistrationBackgroundFetchAPI::~ServiceWorkerRegistrationBackgroundFetchAPI):
(WebCore::ServiceWorkerRegistrationBackgroundFetchAPI::backgroundFetch):
(WebCore::ServiceWorkerRegistrationBackgroundFetchAPI::backgroundFetchIfCreated):
(WebCore::ServiceWorkerRegistrationBackgroundFetchAPI::backgroundFetchManager):
(WebCore::ServiceWorkerRegistrationBackgroundFetchAPI::from):
(WebCore::ServiceWorkerRegistrationBackgroundFetchAPI::supplementName):
* Source/WebCore/workers/service/background-fetch/ServiceWorkerRegistrationBackgroundFetchAPI.h: Added.

Canonical link: <a href="https://commits.webkit.org/260532@main">https://commits.webkit.org/260532@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2aeaec1724d025aab78ab4991c3cd6666dae42d9

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/108629 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/17730 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/41483 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/117739 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/117940 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/112512 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/19181 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/9007 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/100861 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/114396 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/14379 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/97612 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/42352 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/96355 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/29250 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/84082 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/97762 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/10538 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/30600 "Passed tests") | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/38/builds/98597 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/8654 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/11296 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/7514 "Passed tests") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/38/builds/98597 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/16687 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/50196 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/106189 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7279 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/12884 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/26318 "Passed tests") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on LayoutTests/imported/w3c/web-platform-tests/background-fetch/interfaces.worker-expected.txt:
```diff
@@ -1,51 +1,8 @@
+    RenderBody {BODY} at (8,13) size 784x15
+      RenderBlock {PRE} at (0,0) size 784x15
+        RenderText {#text} at (0,0) size 328x15
+          text run at (0,0) width 328: "{\"error\": {\"code\": 404, \"message\": \"404\"}}"
```
The test is not running as expected there. We should probably figure this out. Might be missing a sub resource. Or maybe this file should be dropped completely since it seems LayoutTests/imported/w3c/web-platform-tests/background-fetch/interfaces.worker.html got deleted in this patch.

youennf replied:
It should be removed as well, WPT is no longer having these tests.

cdumez commented on Source/JavaScriptCore/runtime/IntlListFormat.h:
```diff
@@ -30,7 +30,6 @@
 
 #if !defined(HAVE_ICU_U_LIST_FORMATTER)
 #if U_ICU_VERSION_MAJOR_NUM >= 67 || (U_ICU_VERSION_MAJOR_NUM >= 66 && USE(APPLE_INTERNAL_SDK))
-#define HAVE_ICU_U_LIST_FORMATTER 1
```
I don't think you meant to include this?

youennf replied:
right

cdumez commented on Source/WebCore/workers/service/background-fetch/BackgroundFetchEvent.cpp:
```diff
@@ -0,0 +1,63 @@
+
+RefPtr<BackgroundFetchRegistration> BackgroundFetchEvent::registration() const
+{
+    fprintf(stderr, "BackgroundFetchEvent::registration\n");
```
Should probably be dropped.

youennf replied:
right

cdumez commented on Source/WebCore/workers/service/background-fetch/BackgroundFetchEvent.idl:
```diff
@@ -0,0 +1,33 @@
+    EnabledBySetting=ServiceWorkersEnabled&BackgroundFetchAPIEnabled,
+    Exposed=ServiceWorker
+] interface BackgroundFetchEvent : ExtendableEvent {
+  constructor([AtomString] DOMString type, BackgroundFetchEventInit init);
```
Wrong indentation.

youennf replied:
Specs seem to use use 2 space indents for WebIDL.
Fixed.

cdumez commented on Source/WebCore/workers/service/background-fetch/BackgroundFetchEventInit.idl:
```diff
@@ -0,0 +1,30 @@
+[
+    Conditional=SERVICE_WORKER,
+] dictionary BackgroundFetchEventInit : ExtendableEventInit {
+  required BackgroundFetchRegistration registration;
```
Wrong indentation.

cdumez commented on Source/WebCore/workers/service/background-fetch/BackgroundFetchFailureReason.h:
```diff
@@ -0,0 +1,61 @@
+
+namespace WTF {
+
+template<> struct EnumTraits<WebCore::BackgroundFetchFailureReason> {
```
In modern code, I believe we should be using WebCoreArgumentCodersCocoa.serialization.in instead.

youennf replied:
Let's keep it there for now, I am unsure whether we will persist such data, in which case it might be convenient to keep it here.

cdumez commented on Source/WebCore/workers/service/background-fetch/BackgroundFetchInformation.h:
```diff
@@ -0,0 +1,54 @@
+    bool recordsAvailable { true };
+};
+
+
```
Extra blank line.

youennf replied:
ok

cdumez commented on Source/WebCore/workers/service/background-fetch/BackgroundFetchRegistration.h:
```diff
@@ -0,0 +1,94 @@
+struct CacheQueryOptions;
+class FetchRequest;
+
+class BackgroundFetchRegistration : public RefCounted<BackgroundFetchRegistration>, public EventTarget, public ActiveDOMObject {
```
final?

youennf replied:
right

cdumez commented on Source/WebCore/workers/service/background-fetch/BackgroundFetchRegistration.h:
```diff
@@ -0,0 +1,94 @@
+
+struct BackgroundFetchInformation;
+class BackgroundFetchRecord;
+struct BackgroundFetchRecordInformation;
```
We shouldn't mix struct and class forward declarations. I believe our style is first class then struct.

youennf replied:
ok

cdumez commented on Source/WebCore/workers/service/background-fetch/BackgroundFetchRegistration.h:
```diff
@@ -0,0 +1,94 @@
+    static Ref<BackgroundFetchRegistration> create(ScriptExecutionContext&, BackgroundFetchInformation&&);
+    ~BackgroundFetchRegistration();
+
+    String id() const { return m_information.identifier; }
```
Should return a `const String&`

youennf replied:
right

cdumez commented on Source/WebCore/workers/service/background-fetch/BackgroundFetchRegistration.h:
```diff
@@ -0,0 +1,94 @@
+    uint64_t downloaded() const { return m_information.downloaded; }
+    BackgroundFetchResult result() { return m_information.result; }
+    BackgroundFetchFailureReason failureReason() { return m_information.failureReason; }
+    bool recordsAvailable() { return m_information.recordsAvailable; }
```
Should be marked as const. Also, this would look much better with a prefix like "areRecordsAvailable".

youennf replied:
Added const, I kept recordsAvailable since it will probably only be used by binding code

cdumez commented on Source/WebCore/workers/service/background-fetch/BackgroundFetchResult.h:
```diff
@@ -0,0 +1,53 @@
+
+namespace WTF {
+
+template<> struct EnumTraits<WebCore::BackgroundFetchResult> {
```
Should use the new serialization.in file.

cdumez commented on Source/WebCore/workers/service/background-fetch/BackgroundFetchUIOptions.h:
```diff
@@ -0,0 +1,42 @@
+
+struct BackgroundFetchUIOptions {
+    Vector<ImageResource> icons;
+    String title;
```
We should include the header for String.

youennf replied:
ok

youennf replied:
Actually this is not needed, ImageResource already includes it.

cdumez commented on Source/WebCore/workers/service/background-fetch/BackgroundFetchUIOptions.idl:
```diff
@@ -0,0 +1,31 @@
+[
+    Conditional=SERVICE_WORKER,
+] dictionary BackgroundFetchUIOptions {
+  sequence<ImageResource> icons;
```
Wrong indentation.

cdumez commented on Source/WebCore/workers/service/background-fetch/BackgroundFetchUpdateUIEvent.cpp:
```diff
@@ -0,0 +1,62 @@
+
+WTF_MAKE_ISO_ALLOCATED_IMPL(BackgroundFetchUpdateUIEvent);
+
+Ref<BackgroundFetchUpdateUIEvent> BackgroundFetchUpdateUIEvent::create(const AtomString& type, Init&& init, IsTrusted  isTrusted)
```
extra space before `isTrusted`

youennf replied:
ok

cdumez commented on Source/WebCore/workers/service/background-fetch/BackgroundFetchUpdateUIEvent.h:
```diff
@@ -0,0 +1,55 @@
+class BackgroundFetchUpdateUIEvent final : public BackgroundFetchEvent {
+    WTF_MAKE_ISO_ALLOCATED(BackgroundFetchUpdateUIEvent);
+public:
+    ~BackgroundFetchUpdateUIEvent();
```
Would be nice if this was below the create() function, like we usually do.

youennf replied:
ok

cdumez commented on Source/WebCore/workers/service/background-fetch/ServiceWorkerRegistrationBackgroundFetchAPI.cpp:
```diff
@@ -0,0 +1,82 @@
+    return *m_backgroundFetchManager;
+}
+
+ServiceWorkerRegistrationBackgroundFetchAPI* ServiceWorkerRegistrationBackgroundFetchAPI::from(ServiceWorkerRegistration& serviceWorkerRegistration)
```
Should probably return a reference instead of a raw pointer.

youennf replied:
right

cdumez reviewed (approved):
LGTM with changes.

youennf committed:
Add background fetch WebIDL
https://bugs.webkit.org/show_bug.cgi?id=251980
rdar://problem/105213321

Reviewed by Chris Dumez.

Introduce WebIDL, experimental flag and minimal implementation.

Covered by rebased tests.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/background-fetch/abort.https.window-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/background-fetch/content-security-policy.https.window-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/background-fetch/fetch-uploads.https.window-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/background-fetch/fetch.https.window-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/background-fetch/get-ids.https.window-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/background-fetch/get.https.window-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/background-fetch/idlharness.https.any-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/background-fetch/idlharness.https.any.serviceworker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/background-fetch/idlharness.https.any.sharedworker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/background-fetch/idlharness.https.any.worker-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/background-fetch/interfaces.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/background-fetch/match.https.window-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/background-fetch/mixed-content-and-allowed-schemes.https.window-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/background-fetch/port-blocking.https.window-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/background-fetch/update-ui.https.window-expected.txt: Added.
* Source/JavaScriptCore/runtime/IntlListFormat.h:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/fetch/FetchHeaders.h:
* Source/WebCore/Modules/fetch/FetchHeadersGuard.h: Added.
* Source/WebCore/Modules/fetch/FetchResponseBodyLoader.h: Added.
(WebCore::FetchResponseBodyLoader::takeConsumeDataCallback):
(WebCore::FetchResponseBodyLoader::hasConsumeDataCallback const):
(WebCore::FetchResponseBodyLoader::consumeDataCallback):
(WebCore::FetchResponseBodyLoader::consumeDataByChunk):
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/dom/EventNames.h:
* Source/WebCore/dom/EventNames.in:
* Source/WebCore/dom/EventTargetFactory.in:
* Source/WebCore/workers/service/background-fetch/BackgroundFetchEvent.cpp: Added.
(WebCore::BackgroundFetchEvent::create):
(WebCore::BackgroundFetchEvent::BackgroundFetchEvent):
(WebCore::BackgroundFetchEvent::~BackgroundFetchEvent):
(WebCore::BackgroundFetchEvent::registration const):
* Source/WebCore/workers/service/background-fetch/BackgroundFetchEvent.h: Added.
* Source/WebCore/workers/service/background-fetch/BackgroundFetchEvent.idl: Added.
* Source/WebCore/workers/service/background-fetch/BackgroundFetchEventInit.h: Added.
* Source/WebCore/workers/service/background-fetch/BackgroundFetchEventInit.idl: Added.
* Source/WebCore/workers/service/background-fetch/BackgroundFetchFailureReason.h: Added.
* Source/WebCore/workers/service/background-fetch/BackgroundFetchFailureReason.idl: Added.
* Source/WebCore/workers/service/background-fetch/BackgroundFetchInformation.h: Added.
(WebCore::BackgroundFetchInformation::isolatedCopy const):
(WebCore::BackgroundFetchInformation::isolatedCopy):
* Source/WebCore/workers/service/background-fetch/BackgroundFetchManager.cpp: Added.
(WebCore::BackgroundFetchManager::BackgroundFetchManager):
(WebCore::BackgroundFetchManager::~BackgroundFetchManager):
(WebCore::BackgroundFetchManager::fetch):
(WebCore::BackgroundFetchManager::get):
(WebCore::BackgroundFetchManager::getIds):
* Source/WebCore/workers/service/background-fetch/BackgroundFetchManager.h: Added.
(WebCore::BackgroundFetchManager::create):
* Source/WebCore/workers/service/background-fetch/BackgroundFetchManager.idl: Added.
* Source/WebCore/workers/service/background-fetch/BackgroundFetchOptions.h: Added.
(WebCore::BackgroundFetchOptions::BackgroundFetchOptions):
(WebCore::BackgroundFetchOptions::downloadTotal):
(WebCore::BackgroundFetchOptions::isolatedCopy const):
(WebCore::BackgroundFetchOptions::isolatedCopy):
* Source/WebCore/workers/service/background-fetch/BackgroundFetchOptions.idl: Added.
* Source/WebCore/workers/service/background-fetch/BackgroundFetchRecord.cpp: Added.
(WebCore::BackgroundFetchRecord::BackgroundFetchRecord):
(WebCore::BackgroundFetchRecord::~BackgroundFetchRecord):
(WebCore::BackgroundFetchRecord::request):
(WebCore::BackgroundFetchRecord::settleResponseReadyPromise):
* Source/WebCore/workers/service/background-fetch/BackgroundFetchRecord.h: Added.
(WebCore::BackgroundFetchRecord::create):
(WebCore::BackgroundFetchRecord::responseReady):
* Source/WebCore/workers/service/background-fetch/BackgroundFetchRecord.idl: Added.
* Source/WebCore/workers/service/background-fetch/BackgroundFetchRecordIdentifier.h: Added.
* Source/WebCore/workers/service/background-fetch/BackgroundFetchRecordInformation.h: Added.
(WebCore::BackgroundFetchRecordInformation::isolatedCopy const):
(WebCore::BackgroundFetchRecordInformation::isolatedCopy):
* Source/WebCore/workers/service/background-fetch/BackgroundFetchRegistration.cpp: Added.
(WebCore::BackgroundFetchRegistration::create):
(WebCore::BackgroundFetchRegistration::BackgroundFetchRegistration):
(WebCore::BackgroundFetchRegistration::~BackgroundFetchRegistration):
(WebCore::BackgroundFetchRegistration::abort):
(WebCore::BackgroundFetchRegistration::match):
(WebCore::BackgroundFetchRegistration::matchAll):
(WebCore::BackgroundFetchRegistration::updateInformation):
(WebCore::BackgroundFetchRegistration::activeDOMObjectName const):
(WebCore::BackgroundFetchRegistration::stop):
(WebCore::BackgroundFetchRegistration::virtualHasPendingActivity const):
* Source/WebCore/workers/service/background-fetch/BackgroundFetchRegistration.h: Added.
(WebCore::BackgroundFetchRegistration::id const):
(WebCore::BackgroundFetchRegistration::uploadTotal const):
(WebCore::BackgroundFetchRegistration::uploaded const):
(WebCore::BackgroundFetchRegistration::downloadTotal const):
(WebCore::BackgroundFetchRegistration::downloaded const):
(WebCore::BackgroundFetchRegistration::result):
(WebCore::BackgroundFetchRegistration::failureReason):
(WebCore::BackgroundFetchRegistration::recordsAvailable):
(WebCore::BackgroundFetchRegistration::registrationIdentifier const):
* Source/WebCore/workers/service/background-fetch/BackgroundFetchRegistration.idl: Added.
* Source/WebCore/workers/service/background-fetch/BackgroundFetchRequest.h: Added.
(WebCore::BackgroundFetchRequest::isolatedCopy const):
(WebCore::BackgroundFetchRequest::isolatedCopy):
* Source/WebCore/workers/service/background-fetch/BackgroundFetchResult.h: Added.
* Source/WebCore/workers/service/background-fetch/BackgroundFetchResult.idl: Added.
* Source/WebCore/workers/service/background-fetch/BackgroundFetchUIOptions.h: Added.
* Source/WebCore/workers/service/background-fetch/BackgroundFetchUIOptions.idl: Added.
* Source/WebCore/workers/service/background-fetch/BackgroundFetchUpdateUIEvent.cpp: Added.
(WebCore::BackgroundFetchUpdateUIEvent::create):
(WebCore::BackgroundFetchUpdateUIEvent::BackgroundFetchUpdateUIEvent):
(WebCore::BackgroundFetchUpdateUIEvent::~BackgroundFetchUpdateUIEvent):
(WebCore::BackgroundFetchUpdateUIEvent::updateUI):
* Source/WebCore/workers/service/background-fetch/BackgroundFetchUpdateUIEvent.h: Added.
* Source/WebCore/workers/service/background-fetch/BackgroundFetchUpdateUIEvent.idl: Added.
* Source/WebCore/workers/service/background-fetch/ImageResource.h: Added.
(WebCore::ImageResource::isolatedCopy const):
(WebCore::ImageResource::isolatedCopy):
* Source/WebCore/workers/service/background-fetch/ImageResource.idl: Added.
* Source/WebCore/workers/service/background-fetch/ServiceWorkerGlobalScope+BackgroundFetchAPI.idl: Added.
* Source/WebCore/workers/service/background-fetch/ServiceWorkerRegistration+BackgroundFetchAPI.idl: Added.
* Source/WebCore/workers/service/background-fetch/ServiceWorkerRegistrationBackgroundFetchAPI.cpp: Added.
(WebCore::ServiceWorkerRegistrationBackgroundFetchAPI::ServiceWorkerRegistrationBackgroundFetchAPI):
(WebCore::ServiceWorkerRegistrationBackgroundFetchAPI::~ServiceWorkerRegistrationBackgroundFetchAPI):
(WebCore::ServiceWorkerRegistrationBackgroundFetchAPI::backgroundFetch):
(WebCore::ServiceWorkerRegistrationBackgroundFetchAPI::backgroundFetchIfCreated):
(WebCore::ServiceWorkerRegistrationBackgroundFetchAPI::backgroundFetchManager):
(WebCore::ServiceWorkerRegistrationBackgroundFetchAPI::from):
(WebCore::ServiceWorkerRegistrationBackgroundFetchAPI::supplementName):
* Source/WebCore/workers/service/background-fetch/ServiceWorkerRegistrationBackgroundFetchAPI.h: Added.

Canonical link: https://commits.webkit.org/260532@main
